### PR TITLE
RFC: [JENKINS-29068] Remove user-facing system scripting from core

### PR DESCRIPTION
**Pull request by daniel-beck**
See [JENKINS-29068](https://issues.jenkins-ci.org/browse/JENKINS-29068) (well, sort of).

*This PR is less about the implementation and more about the idea. Let's build a consensus on that first.*

### WAT!?

The Groovy scripting capabilities in Jenkins are a _very_ advanced feature -- so advanced that I think they have no business being in the default installation.

While a flexible solution for power users, it's just too easy for beginner or even intermediate users to shoot themselves in the foot. The presence of this feature also sends the wrong message: It's a hack around deficits of an incomplete solution, and presents a practically insurmountable hurdle in learning even a minimal subset of "Jenkins".

Additionally, even for those with strong knowledge of Jenkins internals, it seems counterproductive to fall back to the script console for everything Jenkins can't do (as opposed to be motivated to address that through actual features or fixes).

Therefore I propose to remove these features from the default install, and move them into a plugin.

(`init.groovy` (etc.) support has been retained deliberately, as it's a completely different use case than than user-facing scripting while Jenkins is running. Likewise advanced – canned – diagnostic features.)

### TODO

This isn't deletion, but rather intended as extraction into a plugin

My current approach would be to drop internal API compatibility, as none of the removed APIs were intended for use – `Jenkins._doScript` is used by a crazy Maven Plugin feature, otherwise I could find no uses of the removed code – but again, this is less about the implementation and more the general idea.

AFAICT we currently have no mechanism to install a plugin on upgrade (i.e. retaining existing users' experience), while not implying a dependency from plugins (which is undesirable in this case IMO).

TBD: Impact on config management tooling. Summoning @rtyler and Job DSL's @daspilker to the discussion.

### Proposed changelog entries

* Remove user-facing Groovy and Jelly scripting features from core: Script Console; `groovy` and `groovysh` CLI commands; and `/script`, `/scriptText`, and `/eval` endpoints.

@jenkinsci/code-reviewers

### CODE REVIEW

**Comment by daspilker** [comment]
This should not have any impact on Job DSL as long as you keep the core dependency on Groovy.

+1 for moving this to a plugin.

File: test/src/test/java/lib/form/ExpandableTextboxTest.java [hunk 1]
```diff
@@ -1,62 +0,0 @@
...
-        assertEquals(jenkins.getSystemMessage(),textareas.item(0).getTextContent());
-    }
-
-    /**
-     * Evaluates the literal Jelly script passed as a parameter as HTML and returns the page.
-     */
-    protected HtmlPage evaluateAsHtml(String jellyScript) throws Exception {
-        HudsonTestCase.WebClient wc = new WebClient();
-        
-        WebRequest req = new WebRequest(wc.createCrumbedUrl("eval"), POST);
```
Comment (jglick) [inline test/src/test/java/lib/form/ExpandableTextboxTest.java]: Probably there is something else using expandable textboxes that could be swapped in here.

**Comment by daniel-beck** [comment]
This is on hold; will go through JEP process unless https://groups.google.com/d/msg/jenkinsci-dev/YfbODkibFI4/nBKI5HfeCwAJ is universally disliked; but so far it looks like I get to at least propose one.

**Comment by jglick** [comment]
FTR this is more about [JENKINS-35514](https://issues.jenkins-ci.org/browse/JENKINS-35514) since it is not about moving Groovy itself.

**Review by Wadeck — COMMENTED** [review]
I like the approach 👍

**Comment by daniel-beck** [comment]
JEPing this.

**Pull request closed without merging**